### PR TITLE
Rearchitect for 2025

### LINKED ISSUE #96: Math expression parsing for numerical entry text boxes
Example: typing `2 + 3` into a number text box should enter the value of `5`.

Basic mathematical syntax, functions, and constants should be supported. `(sin(pi/2)+1*2)^2` will result in `9`.

An example implementation of this is in [HelixNavigator](https://github.com/TripleHelixProgramming/HelixNavigator/blob/master/app/src/main/java/org/team2363/lib/math/ExpressionParser.java).

### LINKED ISSUE #565: Add flip parameter to point-at constraint
Supersedes #502 
>This would allow teams to point with whichever side of their robot is the side they need to use. Eg: you have an intake on the back of your robot but want to point at target with the intake while intaking.

Due to trajectory mirroring, the only usable options are "toward" and "away" so this is just a boolean parameter.

### LINKED ISSUE #566: Use project-wide variables in expression parsing
Using mathjs, allow user to specify custom variables (whose values are also expressions) that can be used in expressions. Have both individual variables and pose variables, grouping of x/y/heading.

### LINKED ISSUE #568: Move path data to individual .traj files
The .traj file would now be read back in to Choreo and be both the input and output format for the app.

### LINKED ISSUE #569: Support non-centered bumpers and non-rectangular module locations

### PR DESCRIPTION
**Pull request by shueja**
* expression parsing instead of number inputs
* new file schema
* redesigned constraints system

Closes #565 
- [X] ui
- [x] trajoptlib

Closes #96 

Closes #566
- [X] file schema
- [X] backend storage of variables

(UI panel to add/edit variables is separate issue)

Closes #568

Closes #569 

Closes #565

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -14,8 +14,8 @@
   "private": true,
   "type": "module",
   "scripts": {
-    "fmtCpp": "wpiformat",
-    "fmtJava": "cd choreolib && ./gradlew spotlessApply",
+    "fmtCpp": "wpiformat -f choreolib && wpiformat -f trajoptlib",
+    "fmtJava": "cd choreolib && bash ./gradlew spotlessApply",
```
Comment (spacey-sooty) [inline package.json:18]: Does this work on Windows?

File: src-tauri/src/main.rs [hunk 2]
```diff
@@ -1,145 +1,19 @@
 // Prevents additional console window on Windows in release, DO NOT REMOVE!!
-#![cfg_attr(not(debug_assertions), windows_subsystem = "windows")]
+//#![cfg_attr(not(debug_assertions), windows_subsystem = "windows")]
```
Comment (spacey-sooty) [inline src-tauri/src/main.rs:2]: Was this intentional?

File: package.json (hunk 1 above)
Comment (shueja) [reply package.json:18]: It worked on my windows. I don't know how to make this work cross-platform where node defaults to cmd.exe on windows and a UNIX-style shell on other platforms.

**Comment by spacey-sooty** [comment]
Closes https://github.com/SleipnirGroup/Choreo/issues/575

**Comment by shueja** [comment]
Closes #576

File: testing/NewPath.traj
Comment (Lewis-Seiden) [inline testing/NewPath.traj:1]: rename to "Tests.traj" or something similar

**Review by Lewis-Seiden — COMMENTED** [review]
didnt go through the typescript changes too thoroughly, but otherwise seems mostly good. a few spots need a little bit of documentation/clarification to me

File: src-tauri/src/document/v2025_0_0.rs [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+}
+#[derive(Debug, Serialize, Deserialize, Clone, Copy)]
+pub struct RobotConfig<T> {
+    pub modules: [Module<T>; 4],
+    pub mass: T,
+    pub inertia: T,
+    pub gearing: T,
+    pub radius: T,
+    /// motor rpm
+    pub vmax: T, // motor rpm
```
[the remark this answers is not in the document]
Comment (shueja) [reply src-tauri/src/document/v2025_0_0.rs:93]: not a fan of rpm as the base unit or as the default expression in the robot config?

[the remark this answers is not in the document]
Comment (shueja) [reply src-tauri/src/document/v2025_0_0.rs:93]: Currently the base unit is what the expression will display as, and default to if you just put in a number

File: src/components/config/WaypointConfigPanel.tsx [hunk 4]
```diff
@@ -123,6 +114,19 @@ class WaypointPanel extends Component<Props, State> {
...
+              padding: "inherit",
+              background: "var(--darker-purple)",
+              borderBottomRightRadius: "8px",
+              fontWeight: "bolder",
+              fontSize: "1em",
+
+            }}
+          >
+            {index + 1}
+          </span> */}
```
[the remark this answers is not in the document]
Comment (shueja) [reply src/components/config/WaypointConfigPanel.tsx:129]: I clean this up in the constraints pr.

File: src/components/config/ExportConfigPanel.tsx [hunk 5]
```diff
@@ -23,16 +25,16 @@ class ExportConfigPanel extends Component<Props, State> {
...
         <Switch
           size="small"
           sx={{ gridColumn: 2 }}
-          checked={this.context.model.document.splitTrajectoriesAtStopPoints}
+          checked={doc.traj.useModuleForces}
           onChange={(e, checked) => {
-            this.context.model.setSplitTrajectoriesAtStopPoints(checked);
-            this.context.clearAllTrajectories();
-            this.context.exportAllTrajectories();
+            doc.setSplitTrajectoriesAtStopPoints(checked);
```
[the remark this answers is not in the document]
Comment (shueja) [reply src/components/config/ExportConfigPanel.tsx]: fixed

**Review by Lewis-Seiden — APPROVED** [review]

**Pull request merged**